### PR TITLE
Introduce ParseFailureException to Positions CSV Loader

### PR DESCRIPTION
**Pull request by akshaishah**

### CODE REVIEW

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/CsvLoaderUtils.java [hunk 1]
```diff
@@ -228,22 +229,34 @@ public static Pair<YearMonth, EtdVariant> parseEtdVariant(CsvRow row, EtdType ty
...
-          throw new ParseFailureException(
-              "Unable to parse ETD variant, date columns conflict, must not  set both expiry day and expiry week");
+          throw new ParseFailureException(FailureReason.INVALID,
+              "Unable to parse ETD variant, date columns conflict, must not set both expiry day and expiry week");
         }
       }
     } else {
       if (day == 0) {
-        throw new ParseFailureException("Unable to parse ETD variant for Flex '{value}', must set expiry day", type);
+        throw new ParseFailureException(
```
Comment (jodastephen) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/CsvLoaderUtils.java:238]: These error messages are consistent at present and we need to not lose information here. Plus, error messages are not sentences - they do not have full stops.

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/CsvLoaderUtils.java [hunk 2]
```diff
@@ -723,4 +737,8 @@ public static String formattedDouble(double value) {
     return Decimal.of(value).toString();
   }
 
+  static ParseFailureException missingFieldException(String field) {
```
Comment (jodastephen) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/CsvLoaderUtils.java:740]: Shouldn't this be in the resolver if it is going to be nicluded

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/PositionCsvLoader.java [hunk 3]
```diff
@@ -378,14 +380,15 @@ private <T extends Position> ValueWithFailures<List<T>> parseFile(CsvIterator cs
               .ifPresent(parsed -> positions.add((T) parsed));
         } else {
           // failed to find the type
-          failures.add(FailureItem.of(
-              FailureReason.PARSING,
-              "CSV position file '{fileName}' contained unknown position type '{type}' at line {lineNumber}",
-              CharSources.extractFileName(charSource),
-              typeRaw,
-              row.lineNumber()));
+          throw new ParseFailureException(FailureReason.UNSUPPORTED, "Unknown '{}', '{}'", POSITION_TYPE_FIELD, typeRaw);
```
Comment (jodastephen) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/PositionCsvLoader.java:383]: We need to keep the detail here. While I understand it is in the attributes, we can't rely on that being used, so it must also be in the message.

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/CsvLoaderUtils.java (hunk 1 above)
Comment (akshaishah) [reply modules/loader/src/main/java/com/opengamma/strata/loader/csv/CsvLoaderUtils.java:238]: Have removed the full stops on all of these error messages as discussed.

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/CsvLoaderUtils.java (hunk 2 above)
Comment (akshaishah) [reply modules/loader/src/main/java/com/opengamma/strata/loader/csv/CsvLoaderUtils.java]: I included it here so we can use it in other parsers within and outside of Strata. Happy to keep it within the resolver if you think that is better.

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/PositionCsvLoader.java (hunk 3 above)
Comment (akshaishah) [reply modules/loader/src/main/java/com/opengamma/strata/loader/csv/PositionCsvLoader.java]: Have ensured all of the existing detail now remains

**Review by jodastephen — CHANGES_REQUESTED** [review]

File: modules/collect/src/main/java/com/opengamma/strata/collect/result/FailureReason.java [hunk 4]
```diff
@@ -78,6 +78,18 @@ public enum FailureReason implements NamedEnum {
    * This is used to indicate that a calculation failed.
    */
   CALCULATION_FAILED,
+  /**
+   * The operation failed because a field was missing.
+   * <p>
+   * This is used to indicate that a field was not provided.
+   */
+  FIELD_MISSING,
```
[the remark this answers is not in the document]
Comment (akshaishah) [reply modules/collect/src/main/java/com/opengamma/strata/collect/result/FailureReason.java]: I've reverted these back to `FailureReason.PARSING`. I am going to see if we can work without the sub-attributes for now. Will raise a follow up PR if we do need to add them later.

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/CsvLoaderUtils.java [hunk 5]
```diff
@@ -228,22 +229,34 @@ public static Pair<YearMonth, EtdVariant> parseEtdVariant(CsvRow row, EtdType ty
         if (week == 0) {
           return Pair.of(yearMonth, EtdVariant.ofDaily(day));
         } else {
-          throw new ParseFailureException(
-              "Unable to parse ETD variant, date columns conflict, must not  set both expiry day and expiry week");
+          throw new ParseFailureException(FailureReason.INVALID,
```
Comment (jodastephen) [inline modules/loader/src/main/java/com/opengamma/strata/loader/csv/CsvLoaderUtils.java:232]:
```suggestion
          throw new ParseFailureException(FailureReason.PARSING,
```

**Review by jodastephen — CHANGES_REQUESTED** [review]

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/CsvLoaderUtils.java [hunk 6]
```diff
@@ -228,22 +229,34 @@ public static Pair<YearMonth, EtdVariant> parseEtdVariant(CsvRow row, EtdType ty
...
+            SETTLEMENT_TYPE_FIELD);
       }
       if (type == EtdType.FUTURE) {
         return Pair.of(yearMonth, EtdVariant.ofFlexFuture(day, settleTypeOpt.get()));
       } else {
         if (!optionTypeOpt.isPresent()) {
-          throw new ParseFailureException("Unable to parse ETD variant for Flex Option, must set option type", type);
+          throw new ParseFailureException(
+              FailureReason.PARSING,
+              "'{field}' is empty. Must be set when '{}' is provided as this denotes a Flex Option",
```
[the remark this answers is not in the document]
Comment (akshaishah) [reply modules/loader/src/main/java/com/opengamma/strata/loader/csv/CsvLoaderUtils.java:257]: Apologies thought I had fixed all of these in my previous PR - not sure what happened there... have sorted now

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/PositionCsvLoader.java [hunk 7]
```diff
@@ -378,14 +381,39 @@ private <T extends Position> ValueWithFailures<List<T>> parseFile(CsvIterator cs
               .ifPresent(parsed -> positions.add((T) parsed));
         } else {
           // failed to find the type
-          failures.add(FailureItem.of(
-              FailureReason.PARSING,
+          FailureItem failureItem = FailureItem.of(
+              FailureReason.UNSUPPORTED,
```
[the remark this answers is not in the document]
Comment (akshaishah) [reply modules/loader/src/main/java/com/opengamma/strata/loader/csv/PositionCsvLoader.java]: Apologies I completely missed these.. have sorted now

File: modules/collect/src/main/java/com/opengamma/strata/collect/result/ParseFailureException.java [hunk 8]
```diff
@@ -67,6 +67,44 @@ public ParseFailureException(Throwable cause, String messageTemplate, Object...
...
+   * If there are too many arguments, then the excess arguments are appended to the
+   * end of the message. No attempt is made to format the arguments.
+   * See {@link Messages#formatWithAttributes(String, Object...)} for more details.
+   *
+   * @param cause  the cause
+   * @param reason the reason for the failure
+   * @param messageTemplate  a message explaining the failure, not empty, uses "{argName}" for inserting {@code messageArgs}
+   * @param messageArgs  the arguments for the message
+   */
+  public ParseFailureException(Throwable cause, FailureReason reason, String messageTemplate, Object... messageArgs) {
```
[the remark this answers is not in the document]
Comment (akshaishah) [reply modules/collect/src/main/java/com/opengamma/strata/collect/result/ParseFailureException.java]: Yep - removed

File: modules/loader/src/main/java/com/opengamma/strata/loader/csv/PositionCsvInfoResolver.java [hunk 9]
```diff
@@ -273,8 +276,10 @@ public default Position parseEtdOptionPosition(CsvRow row, PositionInfo info) {
     EtdContractSpec contract = parseEtdContractSpec(row, EtdType.OPTION);
     Pair<YearMonth, EtdVariant> variant = CsvLoaderUtils.parseEtdVariant(row, EtdType.OPTION);
     int version = row.findValue(VERSION_FIELD).map(Integer::parseInt).orElse(DEFAULT_OPTION_VERSION_NUMBER);
-    PutCall putCall = row.getValue(PUT_CALL_FIELD, LoaderUtils::parsePutCall);
-    double strikePrice = row.getValue(EXERCISE_PRICE_FIELD, LoaderUtils::parseDouble);
+    PutCall putCall = row.findValue(PUT_CALL_FIELD, LoaderUtils::parsePutCall)
```
[the remark this answers is not in the document]
Comment (akshaishah) [reply modules/loader/src/main/java/com/opengamma/strata/loader/csv/PositionCsvInfoResolver.java:279]: Yep good idea - have done so

File: modules/collect/src/main/java/com/opengamma/strata/collect/io/CsvRow.java [hunk 10]
```diff
@@ -155,7 +156,7 @@ public String field(int index) {
   public String getField(String header) {
     Integer index = findIndex(header);
     if (index == null) {
-      throw new IllegalArgumentException("Header not found: '" + header + "'");
+      throw new ParseFailureException("Header not found: '{header}'", header);
```
Comment (jodastephen) [inline modules/collect/src/main/java/com/opengamma/strata/collect/io/CsvRow.java:159]: Just need to update the Javadoc exception for this method and `getField(String header, Function<String, T> postProcessor)`

**Review by jodastephen — CHANGES_REQUESTED** [review]

File: modules/collect/src/main/java/com/opengamma/strata/collect/io/CsvRow.java (hunk 10 above)
Comment (akshaishah) [reply modules/collect/src/main/java/com/opengamma/strata/collect/io/CsvRow.java:159]: Ah good spot - done

**Review by jodastephen — APPROVED** [review]

**Pull request merged**